### PR TITLE
Added staking and voucher contracts

### PR DESCRIPTION
**Pull request by hoongun**

### CODE REVIEW

File: contracts/staking/StakingMain.sol [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+        // check staker balance
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(msg.sender);
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        // append a new stake
+        _pushStake(msg.sender, block.timestamp, votesNumber);
+
+        // TODO: change IERC20 to IVote
+        IERC20(_config.voteAddress).transferFrom(msg.sender, address(this), votesNumber);
```
Comment (BaldyAsh) [inline contracts/staking/StakingMain.sol:41]: The best practice is to make all external calls (even to trusted contracts) after all state changes and events emitting

File: contracts/staking/StakingMain.sol [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            votesNumber += stake.amount;
+
+            _popStake(msg.sender);
+        }
+
+        // ???
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(address(this));
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        IERC20(_config.voteAddress).transfer(msg.sender, votesNumber);
```
Comment (BaldyAsh) [inline contracts/staking/StakingMain.sol:64]: same

File: contracts/staking/StakingMain.sol [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        emit VotesUnstaked(msg.sender, votesNumber);
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(msg.sender, vouchersNumber);
+        emit VouchersMinted(msg.sender, vouchersNumber);
```
Comment (BaldyAsh) [inline contracts/staking/StakingMain.sol:70]: same

File: contracts/staking/StakingMain.sol [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    function withdrawVouchersForAll() external override isOwner {
```
Comment (BaldyAsh) [inline contracts/staking/StakingMain.sol:73]: May run out of gas. Better set some constraints like "start index" & "length"

File: contracts/staking/StakingMain.sol (hunk 1 above)
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: The best practice is to make all external calls (even to trusted contracts) after all state changes and events emitting

File: contracts/staking/StakingMain.sol (hunk 2 above)
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: same

File: contracts/staking/StakingMain.sol [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        // ???
+        uint256 votesBalance = IERC20(_config.voteAddress).balanceOf(address(this));
+        if (votesBalance < votesNumber) revert StakingErrors.NotEnoughVotes();
+
+        IERC20(_config.voteAddress).transfer(msg.sender, votesNumber);
+        emit VotesUnstaked(msg.sender, votesNumber);
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(msg.sender, vouchersNumber);
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol:69]: same

File: contracts/staking/StakingMain.sol (hunk 4 above)
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: May run out of gas. Better set some constraints like "start index" & "length"

File: contracts/staking/StakingMain.sol [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+        // append a new stake
+        _pushStake(msg.sender, block.timestamp, votesNumber);
+
+        // TODO: change IERC20 to IVote
+        IERC20(_config.voteAddress).transferFrom(msg.sender, address(this), votesNumber);
+        emit VotesStaked(block.timestamp, votesNumber);
+    }
+
+    function unstakeVotes() external override {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: May run out of gas. Better set some constraints like "start index" & "length" for stakes

File: contracts/staking/StakingMain.sol [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    function setThreshold(uint256 threshold) external override isOwner {
+        _votesThreshold = threshold;
+    }
+
+    function getThreshold() external view override returns (uint256) {
+        return _votesThreshold;
+    }
+
+    function getVotesAmount() external view override returns (uint256) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Better set some constraints like "start index" & "length"

File: contracts/staking/StakingMain.sol [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 voteNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
+            voteNumber += stakes[i].amount;
+        }
+        return voteNumber;
+    }
+
+    function getVouchersAmount() external view override returns (uint256) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Better set some constraints like "start index" & "length"

File: contracts/staking/StakingMain.sol [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        emit VotesUnstaked(msg.sender, votesNumber);
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(msg.sender, vouchersNumber);
+        emit VouchersMinted(msg.sender, vouchersNumber);
+    }
+
+    function withdrawVouchersForAll() external override isOwner {
+        for (uint256 i = 0; i < _stakers.length; i++) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption

File: contracts/staking/StakingMain.sol [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    function getThreshold() external view override returns (uint256) {
+        return _votesThreshold;
+    }
+
+    function getVotesAmount() external view override returns (uint256) {
+        Stake[] memory stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 voteNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption

File: contracts/staking/StakingMain.sol [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        return voteNumber;
+    }
+
+    function getVouchersAmount() external view override returns (uint256) {
+        Stake[] memory stakes = _stakes[msg.sender];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 vouchersNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption

File: contracts/staking/StakingMain.sol [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        return vouchersNumber;
+    }
+
+    // TODO: during withdrawal we should merge multiple stakes to the only one. it will free storage
+    function withdrawVouchers(address staker) public override isOwner {
+        Stake[] storage stakes = _stakes[staker];
+        if (stakes.length == 0) revert StakingErrors.StakerNotFound();
+
+        uint256 vouchersNumber;
+        for (uint256 i = 0; i < stakes.length; i++) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: Put length in front of the loop "uint length = _stakers.length;" and use it as a higher bound in it. Better for gas consumption

File: contracts/staking/StakingMain.sol [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            vouchersNumber += _calculateVouchers(stakes[i]);
+
+            // reset staking time
+            stakes[i].startDate = block.timestamp;
+        }
+
+        if (vouchersNumber == 0) return;
+
+        IVoucher(_config.voucherAddress).mint(staker, vouchersNumber);
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: .

File: contracts/staking/StakingMain.sol [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+        // remove a stake
+        Stake memory stake = _stakes[staker][_stakes[staker].length - 1];
+        _stakes[staker].pop();
+        emit StakeRemoved(msg.sender, stake.startDate, stake.amount);
+
+        // remove a staker
+        if (_stakes[staker].length == 0) {
+            delete _stakes[staker];
+
+            for (uint256 i = 0; i < _stakers.length; i++) {
```
Comment (zkToha) [inline contracts/staking/StakingMain.sol]: .

**Review by zkToha — CHANGES_REQUESTED** [review]

**Review by zkToha — APPROVED** [review]

**Pull request merged**